From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Apehum <apehumchik@gmail.com>
Date: Tue, 5 Apr 2022 13:37:10 +0800
Subject: [PATCH] Fix Lectern Quick Move


diff --git a/src/main/java/net/minecraft/world/inventory/LecternMenu.java b/src/main/java/net/minecraft/world/inventory/LecternMenu.java
index ff79925bc6437222f9ceb133e21bbc0600cc74ed..aaf3e525bf8f32914e9621345433ab05ec4909d5 100644
--- a/src/main/java/net/minecraft/world/inventory/LecternMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/LecternMenu.java
@@ -61,6 +61,13 @@ public class LecternMenu extends AbstractContainerMenu {
         this.player = (Player) playerinventory.player.getBukkitEntity(); // CraftBukkit
     }
 
+    // Matter start
+    @Override
+    public ItemStack quickMoveStack(net.minecraft.world.entity.player.Player player, int index) {
+        return ItemStack.EMPTY;
+    }
+    // Matter end
+
     @Override
     public boolean clickMenuButton(net.minecraft.world.entity.player.Player player, int id) {
         int j;
